sweepbatcher/sweep_batcher.go

```diff
@@ -0,0 +1,654 @@
+	store BatcherStore, swapStore loopdb.SwapStore) *Batcher {
+
+	return &Batcher{
+		batches:          make(map[int32]*batch),
+		sweepReqs:        make(chan SweepRequest),
+		errChan:          make(chan error, 1),
+		completeChan:     make(chan int32),
+		quit:             make(chan struct{}),
+		wallet:           wallet,
+		chainNotifier:    chainNotifier,
+		signerClient:     signerClient,
+		musig2ServerSign: musig2ServerSigner,
+		VerifySchnorrSig: verifySchnorrSig,
+		chainParams:      chainparams,
+		store:            store,
+		swapStore:        swapStore,
+	}
+}
+
+// Run starts the batcher and processes incoming sweep requests.
+func (b *Batcher) Run(ctx context.Context) error {
+	runCtx, cancel := context.WithCancel(ctx)
+	defer func() {
+		cancel()
+
+		for _, batch := range b.batches {
+			batch.Wait()
+		}
+
+		b.wg.Wait()
+	}()
+
+	// First we fetch all the batches that are not in a confirmed state from
+	// the database. We will then resume the execution of these batches.
+	batches, err := b.FetchUnconfirmedBatches(runCtx)
+	if err != nil {
+		return err
+	}
+
+	for _, batch := range batches {
+		err := b.spinUpBatchFromDB(runCtx, batch)
+		if err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case sweepReq := <-b.sweepReqs:
+			sweep, err := b.fetchSweep(runCtx, sweepReq)
+			if err != nil {
+				return err
+			}
+
+			err = b.handleSweep(runCtx, sweep, sweepReq.Notifier)
+			if err != nil {
+				return err
+			}
+
+		case id := <-b.completeChan:
+			delete(b.batches, id)
+
+		case err := <-b.errChan:
+			return err
+
+		case <-runCtx.Done():
+			return runCtx.Err()
+		}
+	}
+}
+
+// AddSweep adds a sweep request to the batcher for handling. This will either
+// place the sweep in an existing batch or create a new one.
+func (b *Batcher) AddSweep(sweepReq *SweepRequest) error {
+	select {
+	case b.sweepReqs <- *sweepReq:
+		return nil
+
+	case <-b.quit:
+		return ErrBatcherShuttingDown
+	}
+}
+
+// handleSweep handles a sweep request by either placing it in an existing
+// batch, or by spinning up a new batch for it.
+func (b *Batcher) handleSweep(ctx context.Context, sweep *sweep,
+	notifier *SpendNotifier) error {
+
+	completed, err := b.store.GetSweepStatus(ctx, sweep.swapHash)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Batcher handling sweep %x, completed=%v", sweep.swapHash[:6],
+		completed)
+
+	// If the sweep has already been completed in a confirmed batch then we
+	// can't attach its notifier to the batch as that is no longer running.
+	// Instead we directly detect and return the spend here.
+	if completed && *notifier != (SpendNotifier{}) {
+		go b.monitorSpendAndNotify(ctx, sweep, notifier)
+		return nil
+	}
+
+	sweep.notifier = notifier
+
+	// Check if the sweep is already in a batch. If that is the case, we
+	// provide the sweep to that batch and return.
+	for _, batch := range b.batches {
+		if batch.sweepExists(sweep.swapHash) {
+			_, err := batch.addSweep(ctx, sweep)
```

Comment:
nit: consider adding an extra check here that if we didn't error out we accepted the sweep.